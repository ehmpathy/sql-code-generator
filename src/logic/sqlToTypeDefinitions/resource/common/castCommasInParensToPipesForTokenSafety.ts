export const castCommasInParensToPipesForTokenSafety = ({ sql }: { sql: string }) => {
  // track state of whether we are in parenthesizes or not
  let areInsideParensDepth = 0; // init as 0, since not inside parens
  const areInsideParens = () => areInsideParensDepth > 0; // whenever areInsideParensDepth > 0, we are inside parens

  // init object to build up the "casted" string
  let castedString = '';

  // loop through each char and update state / replace char as needed
  [...sql].forEach((char) => {
    // 1. manage "areInsideParens" state
    if (char === '(') {
      areInsideParensDepth += 1; // one parens deeper!
    }
    if (char === ')') {
      if (!areInsideParens) throw new Error('exiting parenthesizes without having opened them; not supported'); // fail fast for unsupported case
      areInsideParensDepth -= 1; // mark that we've dropped off one parens now
    }

    // 2. cast the char, considering whether we're in parens or not
    const castedChar = areInsideParens() && char === ',' ? '|' : char; // cast "," if we are inside parens

    // 3. add casted char to string
    castedString += castedChar;
  });

  // return the casted string
  return castedString;
};
