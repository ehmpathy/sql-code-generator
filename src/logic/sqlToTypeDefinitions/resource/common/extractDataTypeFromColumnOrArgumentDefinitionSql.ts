import { DataType } from '../../../../model';

// https://dev.mysql.com/doc/refman/8.0/en/string-types.html
const mysqlStringTypes = [
  'CHAR',
  'VARCHAR',
  'TINYTEXT',
  'TEXT',
  'MEDIUMTEXT',
  'LONGTEXT',
  'ENUM', // note: we treat enums as strings; TODO: https://github.com/uladkasach/sql-code-generator/issues/1
];

// https://www.postgresql.org/docs/9.5/datatype-character.html
const postgresStringTypes = [
  'CHARACTER',
  'CHAR',
  'CHARACTER VARYING',
  'VARCHAR',
  'TEXT',
  'UUID', // this too: https://www.postgresql.org/docs/9.5/datatype-uuid.html
];

// combine the types so we can search over the union
const dbStringTypes = [...mysqlStringTypes, ...postgresStringTypes];

// https://dev.mysql.com/doc/refman/8.0/en/numeric-types.html
const mysqlNumberTypes = [
  'INTEGER',
  'INT',
  'SMALLINT',
  'TINYINT',
  'MEDIUMINT',
  'BIGINT',
  'DECIMAL',
  'NUMERIC',
  'FLOAT',
  'DOUBLE',
];

// https://www.postgresql.org/docs/9.5/datatype-numeric.html
const postgresNumberTypes = [
  'SMALLINT',
  'INT2',
  'INTEGER',
  'INT',
  'INT4',
  'BIGINT',
  'INT8',
  'DECIMAL',
  'NUMERIC',
  'REAL',
  'DOUBLE PRECISION',
  'FLOAT8',
  'SMALLSERIAL',
  'SERIAL2',
  'SERIAL',
  'SERIAL4',
  'BIGSERIAL',
  'SERIAL8',
];

// combine the types so we can search over the union
const dbNumberTypes = [...mysqlNumberTypes, ...postgresNumberTypes];

// https://dev.mysql.com/doc/refman/8.0/en/date-and-time-types.html
const mysqlDateTypes = ['DATE', 'TIME', 'DATETIME', 'TIMESTAMP', 'YEAR'];

// https://www.postgresql.org/docs/9.5/datatype-datetime.html
const postgresDateTypes = [
  'TIMESTAMP',
  'TIMESTAMPTZ',
  'TIMESTAMP WITH TIMEZONE',
  'DATE',
  'TIME',
  'TIMETZ',
  'TIME WITH TIMEZONE',
];

// combine the types so we can search over the union
const dbDateTypes = [...mysqlDateTypes, ...postgresDateTypes];

// https://dev.mysql.com/doc/refman/8.0/en/binary-varbinary.html
const mysqlBinaryTypes = ['BINARY', 'VARBINARY'];

// https://www.postgresql.org/docs/9.5/datatype-binary.html
const postgresBinaryTypes = ['BYTEA'];

// combine the types so we can search over the union
const dbBinaryTypes = [...mysqlBinaryTypes, ...postgresBinaryTypes];

export const extractDataTypeFromColumnOrArgumentDefinitionSql = ({ sql }: { sql: string }) => {
  // 1. cast to upper
  const sqlUpper = sql.toUpperCase();

  // 2. one by one try to see which datatype is being defined in this definition
  if (dbStringTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.STRING;
  if (dbNumberTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.NUMBER;
  if (dbDateTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.DATE;
  if (dbBinaryTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.BUFFER;
  throw new Error(`could not extract data type from '${sql}'`);
};
