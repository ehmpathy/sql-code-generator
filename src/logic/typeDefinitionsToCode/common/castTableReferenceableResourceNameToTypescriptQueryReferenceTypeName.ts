import { ResourceType } from '../../../model';
import { castResourceNameToTypescriptTypeName } from './castResourceNameToTypescriptTypeName';

export const castTableReferenceableResourceNameToTypescriptQueryReferenceTypeName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  // 0. throw error if not view or table, only view and table are possible table references
  if (![ResourceType.TABLE, ResourceType.VIEW].includes(resourceType)) {
    throw new Error('not a table referenceable query');
  }

  // 1. get the regular typescript type name
  const typescriptTypeName = castResourceNameToTypescriptTypeName({ name, resourceType });

  // 2. return the referenceable name
  if (resourceType === ResourceType.TABLE) return typescriptTypeName.replace(/^SqlTable/, 'SqlTableOrView');
  if (resourceType === ResourceType.VIEW) return typescriptTypeName.replace(/^SqlView/, 'SqlTableOrView');
  throw new Error('unexpected'); // this should never occur, but fail fast
};
