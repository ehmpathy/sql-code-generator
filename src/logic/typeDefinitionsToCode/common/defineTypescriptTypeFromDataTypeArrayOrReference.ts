import {
  DataType,
  TypeDefinitionReference,
  TypeDefinition,
  TypeDefinitionOfQueryTableReference,
} from '../../../domain';
import { defineTypescriptTypeFromDataTypeArray } from './defineTypescriptTypeFromDataTypeArray';
import { defineTypescriptTypeFromReference } from './defineTypescriptTypeFromReference/defineTypescriptTypeFromReference';

export const defineTypescriptTypeFromDataTypeArrayOrReference = ({
  type,
  plural,
  queryTableReferences,
  typeDefinitions,
}: {
  type: DataType[] | TypeDefinitionReference;
  plural: boolean;
  typeDefinitions: TypeDefinition[];
  queryTableReferences: TypeDefinitionOfQueryTableReference[];
}) => {
  // if its a type reference, then use that handler
  if (type instanceof TypeDefinitionReference) {
    const defined = defineTypescriptTypeFromReference({
      reference: type,
      queryTableReferences,
      typeDefinitions,
    });
    if (plural) return `${defined}[]`;
    return defined;
  }

  // else, it must be data array. use that handler
  return defineTypescriptTypeFromDataTypeArray({ type });
};
