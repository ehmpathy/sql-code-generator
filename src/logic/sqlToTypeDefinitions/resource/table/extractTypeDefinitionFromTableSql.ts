import strip from 'sql-strip-comments';

import { castCommasInParensToPipesForTokenSafety } from '../common/castCommasInParensToPipesForTokenSafety';
import { extractTypeDefinitionFromColumnSql } from './extractTypeDefinitionFromColumnSql';
import { TypeDefinitionOfResourceTable } from '../../../../model';

/*
  note:
    - for table and view, just a map of { property: type }
    - for function and procedure, input type, output type
*/

const KEY_OR_CONSTRAINT_LINE_REGEX = /[KEY|CONSTRAINT].*\s\(/gi;

export const extractTypeDefinitionFromTableSql = ({ name, sql }: { name: string; sql: string }) => {
  // 0. strip comments
  const strippedSql: string = strip(sql);

  // 1. drop everything after the first `;` - the table ddl does not have `;` inside. (this will make sure we exclude lines like "create index ..." and etc)
  const sqlBeforeFirstSemiColon = strippedSql.split(';')[0];
  // strippedSql.replace(/CREATE INDEX [\w\d_]+ ON [\w\d_]+ USING [\w\d_]+ \([\w\d_,]+\);/gi, '');

  // 2. grab the insides of the "create" (i.e., 'CREATE TABLE ... (__INSIDES__) ...' => '__INSIDES__')
  const innerSqlAndAfter = sqlBeforeFirstSemiColon
    .split('(')
    .slice(1)
    .join('('); // drop the part before the first '('
  const innerSql = innerSqlAndAfter
    .split(')')
    .slice(0, -1)
    .join(')'); // drop the part after the last ')'

  // 3. cast commas inside of parens into pipes, so that we treat them as unique tokens when splitting "property lines" by comma
  const parenTokenizedInnerSql = castCommasInParensToPipesForTokenSafety({ sql: innerSql });

  // 4. grab definition lines, by splitting out properties by commas
  const lines = parenTokenizedInnerSql.split(',');

  // 5. strip out the lines that are defining keys or constraints
  const columnDefinitions = lines
    .filter((line) => {
      return !new RegExp(KEY_OR_CONSTRAINT_LINE_REGEX, 'gim').test(line); // note, we reinstantiate so as to not share regex object (i.e., state) between calls
    })
    .map((line) => line.trim());

  // 6. get column definition from each property
  const columns = columnDefinitions.map((line) => extractTypeDefinitionFromColumnSql({ sql: line }));

  // 7. return the table definition
  return new TypeDefinitionOfResourceTable({
    name,
    columns,
  });
};
