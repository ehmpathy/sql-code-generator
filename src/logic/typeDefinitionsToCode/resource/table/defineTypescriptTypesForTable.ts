import { TypeDefinitionOfResourceTable, ResourceType } from '../../../../model';
import { castResourceNameToTypescriptTypeName } from '../../common/castResourceNameToTypescriptTypeName';
import { castTableReferenceableResourceNameToTypescriptQueryReferenceTypeName } from '../../common/castTableReferenceableResourceNameToTypescriptQueryReferenceTypeName';

export const defineTypescriptTypesForTable = ({ definition }: { definition: TypeDefinitionOfResourceTable }) => {
  // define column types in typescript format
  const typescriptInterfaceColumnDefinitions = definition.columns.map((column) => {
    return `${column.name}: ${column.type.join(' | ')};`;
  });

  // define interface
  const typescriptTypeName = castResourceNameToTypescriptTypeName({
    name: definition.name,
    resourceType: ResourceType.TABLE,
  });
  const typescriptInterfaceDefinition = `
export interface ${typescriptTypeName} {
  ${typescriptInterfaceColumnDefinitions.join('\n  ')}
}
  `.trim();

  // define a type alias that can be used to reference this generically
  const typescriptQueryReferenceableTypeName = castTableReferenceableResourceNameToTypescriptQueryReferenceTypeName({
    name: definition.name,
    resourceType: ResourceType.TABLE,
  });
  const typescriptReferencableTypeAliasDefinition = `
export type ${typescriptQueryReferenceableTypeName} = ${typescriptTypeName};
  `.trim();

  // return typescript types
  return `
${typescriptInterfaceDefinition}
${typescriptReferencableTypeAliasDefinition}
  `.trim();
};
