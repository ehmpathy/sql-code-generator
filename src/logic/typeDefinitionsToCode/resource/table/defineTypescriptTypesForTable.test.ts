import { extractSqlFromFile } from '../../../common/extractSqlFromFile';
import { extractTypeDefinitionFromTableSql } from '../../../sqlToTypeDefinitions/resource/table/extractTypeDefinitionFromTableSql';
import { defineTypescriptTypesForTable } from './defineTypescriptTypesForTable';

describe('defineTypescriptTypesForTable', () => {
  it('should generate an accurate looking interface for a simple table', async () => {
    const definition = extractTypeDefinitionFromTableSql({
      name: 'image',
      sql: await extractSqlFromFile({ filePath: `${__dirname}/../../../__test_assets__/tables/image.sql` }),
    });
    const code = defineTypescriptTypesForTable({ definition });
    expect(code).toMatchSnapshot();
  });
  it('should generate an accurate looking interface for a more complex table', async () => {
    const definition = extractTypeDefinitionFromTableSql({
      name: 'suggestion_version',
      sql: await extractSqlFromFile({
        filePath: `${__dirname}/../../../__test_assets__/tables/suggestion_version.sql`,
      }),
    });
    const code = defineTypescriptTypesForTable({ definition });
    expect(code).toMatchSnapshot();
  });
});
