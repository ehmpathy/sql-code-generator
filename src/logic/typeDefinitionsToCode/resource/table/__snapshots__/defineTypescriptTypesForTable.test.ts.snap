// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineTypescriptTypesForTable should generate an accurate looking interface for a more complex table 1`] = `
"export interface SqlTableSuggestionVersion {
  id: number;
  suggestion_id: number;
  effective_at: Date;
  created_at: Date;
  status: string;
}
export type SqlTableOrViewSuggestionVersion = SqlTableSuggestionVersion;"
`;

exports[`defineTypescriptTypesForTable should generate an accurate looking interface for a simple table 1`] = `
"export interface SqlTableImage {
  id: number;
  uuid: string;
  created_at: Date;
  url: string;
  caption: string | null;
  credit: string | null;
  alt_text: string | null;
}
export type SqlTableOrViewImage = SqlTableImage;"
`;
